src/libfetchers/git.cc

```diff
@@ -504,15 +504,38 @@ struct GitInputScheme : InputScheme
         std::string name = input.getName();
 
         auto originalRef = input.getRef();
-        auto ref = originalRef ? *originalRef : getDefaultRef(repoInfo);
-        input.attrs.insert_or_assign("ref", ref);
 
-        Path repoDir;
+        Path repoDir ;
+
+        /*
+        The lazy initialization of ref and localRefFile prevents unnecessary
+          network access in cases where those variables are not used.
+        For example, if the `rev` is specified in the input, we don't need to
+          fetch the HEAD ref from the remote repository.
+        */
+        std::optional<std::string> ref;
+        auto getRef = [this, &input, &originalRef, &ref, &repoInfo](){
+            if (!ref) {
+                ref = originalRef ? *originalRef : getDefaultRef(repoInfo);
+                input.attrs.insert_or_assign("ref", *ref);
+            }
+            return *ref;
+        };
+
+        std::optional<std::string> localRefFile;
+        auto getLocalRefFile = [&getRef, &localRefFile, &repoInfo](){
+            if (!localRefFile) {
+                localRefFile = getRef().compare(0, 5, "refs/") == 0
+                    ? repoInfo.url + "/" + getRef()
+                    : repoInfo.url + "/refs/heads/" + getRef();
```

Comment:
Pre-existing issue:

https://github.com/NixOS/nix/pull/10241 suggests that this behavior is a problem. We shouldn't only read refs through libgit2.

Reply:
The test seems to fail here as well. I suppose you could ensure (eagerly) that the cache repo exists, initialize it if needed.